src/librustdoc/passes/collect_intra_doc_links/early.rs

```diff
@@ -1,75 +1,148 @@
+
+use std::collections::hash_map::Entry;
 use std::mem;
-use std::rc::Rc;
-
-type Resolver = Rc<RefCell<interface::BoxedResolver>>;
-// Letting the resolver escape at the end of the function leads to inconsistencies between the
-// crates the TyCtxt sees and the resolver sees (because the resolver could load more crates
-// after escaping). Hopefully `IntraLinkCrateLoader` gets all the crates we need ...
-crate fn load_intra_link_crates(resolver: Resolver, krate: &ast::Crate) -> Resolver {
-    let mut loader = IntraLinkCrateLoader { current_mod: CRATE_DEF_ID, resolver };
-    // `walk_crate` doesn't visit the crate itself for some reason.
+
+crate fn load_intra_link_crates(
+    resolver: &mut Resolver<'_>,
+    krate: &ast::Crate,
+) -> DefIdMap<DefIdSet> {
+    let mut loader = IntraLinkCrateLoader {
+        resolver,
+        current_mod: CRATE_DEF_ID,
+        visited_mods: Default::default(),
+        traits_in_scope: Default::default(),
+    };
+
+    // Overridden `visit_item` below doesn't apply to the crate root,
+    // so we have to visit its attributes and exports separately.
     loader.load_links_in_attrs(&krate.attrs, krate.span);
-    ast::visit::walk_crate(&mut loader, krate);
-    loader.resolver
+    loader.process_exports(CRATE_DEF_ID.to_def_id());
+    visit::walk_crate(&mut loader, krate);
+    loader.add_foreign_traits_in_scope();
+
+    loader.traits_in_scope
 }
 
-struct IntraLinkCrateLoader {
+struct IntraLinkCrateLoader<'r, 'ra> {
+    resolver: &'r mut Resolver<'ra>,
     current_mod: LocalDefId,
-    resolver: Rc<RefCell<interface::BoxedResolver>>,
+    visited_mods: DefIdSet,
+    traits_in_scope: DefIdMap<DefIdSet>,
 }
 
-impl IntraLinkCrateLoader {
-    fn load_links_in_attrs(&mut self, attrs: &[ast::Attribute], span: Span) {
-        use crate::html::markdown::markdown_links;
-        use crate::passes::collect_intra_doc_links::preprocess_link;
+impl IntraLinkCrateLoader<'_, '_> {
+    fn add_traits_in_scope(&mut self, module_id: DefId) {
+        if let Entry::Vacant(entry) = self.traits_in_scope.entry(module_id) {
+            if self.resolver.can_get_module(module_id) {
+                let parent_scope =
+                    ParentScope::module(self.resolver.get_module(module_id), self.resolver);
+                entry.insert(
+                    self.resolver
+                        .traits_in_scope(None, &parent_scope, SyntaxContext::root(), None)
+                        .into_iter()
+                        .map(|candidate| candidate.def_id)
+                        .collect(),
+                );
+            }
+        }
+    }
+
+    /// Add traits in scope for links in impls collected by the `collect-intra-doc-links` pass.
+    /// That pass filters impls using type-based information, but we don't yet have such
+    /// information here, so we just conservatively calculate traits in scope for *all* modules
+    /// having impls in them.
+    fn add_foreign_traits_in_scope(&mut self) {
```

Comment:
nit: since it's wrong for this function to ever be called more than once, do you think it would make sense to inline it into `load_intra_link_crates`?

Reply:
I've left it as is for now, even if `fn add_foreign_traits_in_scope` is called once, it runs its own isolated piece of logic, and `fn load_intra_link_crates` therefore contains only "top level" calls.